Assert in Felix: Unexpected update to LocalEndpoint

Seen on juno FV run 115:

```
2015-12-15 06:37:29,570 [ERROR][18956] calicotest.openstack.basetest 165: ***** Start of errors in logs from /var/log/calico/felix.log on calico-ci03.datcon.co.uk *****

=== LOG 2015-12-15 06:37:18 ERROR [pid 6118] ===
calico.felix.actor 347: _finish_msg_batch failed.
Traceback (most recent call last):
File "/usr/lib/python2.7/dist-packages/calico/felix/actor.py", line 334, in _step
self._finish_msg_batch(batch, results)
File "/usr/lib/python2.7/dist-packages/calico/felix/endpoint.py", line 297, in _finish_msg_batch
(self, self.__dict__)
AssertionError: Unexpected update to LocalEndpoint<IPv6,id=EndpointId<7b0cfe40-3e77-4c2d-a9b7-7bce1bbcc81c>,iface=tap7b0cfe40-3e> ({'_device_in_sync': False, 'rules_mgr': <calico.felix.profilerules.RulesManager object at 0x7fbd89be2310>, '_endpoint_update_pending': False, 'msg_id': None, '_iface_name': u'tap7b0cfe40-3e', 'ref_count': 0, 'combined_id': EndpointId('calico-ci03','openstack','62c7c361-f00c-4e26-8798-a35c534a4bfe','7b0cfe40-3e77-4c2d-a9b7-7bce1bbcc81c'), 'rules_ref_helper': <calico.felix.refcount.RefHelper object at 0x7fbd89075590>, '_chains_programmed': False, 'dispatch_chains': <calico.felix.dispatch.DispatchChains object at 0x7fbd89be23d0>, '_device_is_up': False, '_op_count': 0, 'ip_type': 'IPv6', 'config': <calico.felix.config.Config object at 0x7fbd8a26afd0>, '_last_status': 'down', '_last_scheduled': None, 'qualifier': '7b0cfe40-3e77-4c2d-a9b7-7bce1bbcc81c(IPv6)', 'iptables_updater': <calico.felix.fiptables.IptablesUpdater object at 0x7fbd89be20d0>, 'started': True, '_gevent_loop': <loop at 0x7fbd8a2cc0e0 epoll default pending=0 ref=20 fileno=4>, '_cleaned_up': True, '_event_queue': deque([]), '_added_to_dispatch_chains': True, '_iptables_in_sync': True, '_mac': 'fa:16:3e:c3:2a:50', '_suffix': u'7b0cfe40-3e', '_current_msg': None, 'name': 'LocalEndpoint(7b0cfe40-3e77-4c2d-a9b7-7bce1bbcc81c(IPv6))', 'endpoint': None, '_manager': <weakproxy at 0x7fbd8908c730 to EndpointManager at 0x7fbd89be2410>, 'ref_mgmt_state': 'stopping', '_gevent_hub': <Hub at 0x7fbd8a4f2730 epoll default pending=0 ref=20 fileno=4 resolver=<gevent.resolver_thread.Resolver at 0x7fbd8dc58810 pool=<ThreadPool at 0x7fbd89bb3090 0/1/10>> threadpool=<ThreadPool at 0x7fbd89bb3090 0/1/10>>, '_scheduled': True, '_unreferenced': True, 'greenlet': <Greenlet at 0x7fbd89bdacd0: <bound method LocalEndpoint._loop of <calico.felix.endpoint.LocalEndpoint object at 0x7fbd8aa5a350>>>, 'status_reporter': <calico.felix.fetcd.EtcdStatusReporter object at 0x7fbd89bb38d0>, '_pending_endpoint': None, '_id': EndpointId('calico-ci03','openstack','62c7c361-f00c-4e26-8798-a35c534a4bfe','7b0cfe40-3e77-4c2d-a9b7-7bce1bbcc81c'), '_mac_changed': True}) after being unreferenced

=== LOG 2015-12-15 06:37:18 ERROR [pid 6118] ===
calico.felix.actor 347: _finish_msg_batch failed.
Traceback (most recent call last):
File "/usr/lib/python2.7/dist-packages/calico/felix/actor.py", line 334, in _step
self._finish_msg_batch(batch, results)
File "/usr/lib/python2.7/dist-packages/calico/felix/endpoint.py", line 297, in _finish_msg_batch
(self, self.__dict__)
AssertionError: Unexpected update to LocalEndpoint<IPv6,id=EndpointId<7b0cfe40-3e77-4c2d-a9b7-7bce1bbcc81c>,iface=tap7b0cfe40-3e> ({'_device_in_sync': False, 'rules_mgr': <calico.felix.profilerules.RulesManager object at 0x7fbd89be2310>, '_endpoint_update_pending': False, 'msg_id': None, '_iface_name': u'tap7b0cfe40-3e', 'ref_count': 0, 'combined_id': EndpointId('calico-ci03','openstack','62c7c361-f00c-4e26-8798-a35c534a4bfe','7b0cfe40-3e77-4c2d-a9b7-7bce1bbcc81c'), 'rules_ref_helper': <calico.felix.refcount.RefHelper object at 0x7fbd89075590>, '_chains_programmed': False, 'dispatch_chains': <calico.felix.dispatch.DispatchChains object at 0x7fbd89be23d0>, '_device_is_up': False, '_op_count': 0, 'ip_type': 'IPv6', 'config': <calico.felix.config.Config object at 0x7fbd8a26afd0>, '_last_status': 'down', '_last_scheduled': None, 'qualifier': '7b0cfe40-3e77-4c2d-a9b7-7bce1bbcc81c(IPv6)', 'iptables_updater': <calico.felix.fiptables.IptablesUpdater object at 0x7fbd89be20d0>, 'started': True, '_gevent_loop': <loop at 0x7fbd8a2cc0e0 epoll default pending=0 ref=20 fileno=4>, '_cleaned_up': True, '_event_queue': deque([]), '_added_to_dispatch_chains': True, '_iptables_in_sync': True, '_mac': 'fa:16:3e:c3:2a:50', '_suffix': u'7b0cfe40-3e', '_current_msg': None, 'name': 'LocalEndpoint(7b0cfe40-3e77-4c2d-a9b7-7bce1bbcc81c(IPv6))', 'endpoint': None, '_manager': <weakproxy at 0x7fbd8908c730 to EndpointManager at 0x7fbd89be2410>, 'ref_mgmt_state': 'stopping', '_gevent_hub': <Hub at 0x7fbd8a4f2730 epoll default pending=0 ref=20 fileno=4 resolver=<gevent.resolver_thread.Resolver at 0x7fbd8dc58810 pool=<ThreadPool at 0x7fbd89bb3090 0/1/10>> threadpool=<ThreadPool at 0x7fbd89bb3090 0/1/10>>, '_scheduled': True, '_unreferenced': True, 'greenlet': <Greenlet at 0x7fbd89bdacd0: <bound method LocalEndpoint._loop of <calico.felix.endpoint.LocalEndpoint object at 0x7fbd8aa5a350>>>, 'status_reporter': <calico.felix.fetcd.EtcdStatusReporter object at 0x7fbd89bb38d0>, '_pending_endpoint': None, '_id': EndpointId('calico-ci03','openstack','62c7c361-f00c-4e26-8798-a35c534a4bfe','7b0cfe40-3e77-4c2d-a9b7-7bce1bbcc81c'), '_mac_changed': True}) after being unreferenced


2015-12-15 06:37:29,570 [ERROR][18956] calicotest.openstack.basetest 167: ****** End of errors in logs from /var/log/calico/felix.log on calico-ci03.datcon.co.uk ******
```

--- a/calico/felix/endpoint.py
+++ b/calico/felix/endpoint.py
@@ -268,6 +268,7 @@ class LocalEndpoint(RefCountedActor):
         """
         _log.info("Endpoint %s received interface kick: %s",
                   self.combined_id, iface_up)
+        assert not self._unreferenced, "Interface kick after unreference"
 
         # Use a flag so that we coalesce any duplicate updates in
         # _finish_msg_batch.
@@ -280,6 +281,7 @@ class LocalEndpoint(RefCountedActor):
         Overrides RefCountedActor:on_unreferenced.
         """
         _log.info("%s now unreferenced, cleaning up", self)
+        assert not self._unreferenced, "Duplicate on_unreferenced() call"
 
         # We should be deleted before being unreferenced.
         assert self.endpoint is None or (self._pending_endpoint is None and
@@ -290,12 +292,11 @@ class LocalEndpoint(RefCountedActor):
 
     def _finish_msg_batch(self, batch, results):
         if self._cleaned_up:
-            # We could just ignore this but it suggests that the
-            # EndpointManager is bugged.
-            raise AssertionError(
-                "Unexpected update to %s (%s) after being unreferenced" %
-                (self, self.__dict__)
-            )
+            # This can occur if we get a callback from a profile via the
+            # RefHelper after we've already been deleted.
+            _log.warn("_finish_msg_batch() called after being unreferenced,"
+                      "ignoring.  Batch: %s", batch)
+            return
 
         if self._endpoint_update_pending:
             # Copy the pending update into our data structures.  May work out

--- a/calico/felix/test/test_endpoint.py
+++ b/calico/felix/test/test_endpoint.py
@@ -343,8 +343,7 @@ class TestLocalEndpoint(BaseTestCase):
             self.step_actor(local_ep)
             m_set_routes.assert_called_once_with(ip_type, set(),
                                                  data["name"], None)
-            self.assertRaises(AssertionError,
-                              local_ep._finish_msg_batch, [], [])
+            local_ep._finish_msg_batch([], [])  # Should be ignored
             self.m_manager.on_object_cleanup_complete.assert_called_once_with(
                 local_ep._id,
                 local_ep,

